File: 2021/day1/solution.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;


vector<int> parseDepths(const std::string& filePath) {
    ifstream inputFile(filePath);
    vector<int> lines;
    if (!inputFile) {
        cerr << "Failed to open input file." << endl;
        return lines;
    }
    std::string line;
    while (std::getline(inputFile, line)) {
        lines.push_back(stoll(line));
    }
    inputFile.close();
    return lines;
}

int getIncreaseCount(vector<int> depths) {
    int increases = 0;
    for (int i = 1; i < depths.size(); i++) {
        if (depths[i] > depths[i-1]) {
            increases++;
        }
    }
    return increases;
}

int getSlidingWindowIncreaseCount(vector<int> depths) {
    int increases = 0;
    for (int i = 0; i < depths.size(); i++) {
        if (i < 3) {
            continue;
        }
        int prevWindow = depths[i-3] + depths[i-2] + depths[i-1];
        int currWindow = depths[i-2] + depths[i-1] + depths[i];
        if (currWindow > prevWindow) {
            increases++;
        }
    }
    return increases;
}

int main() {
    string filePath = "input.txt";
    vector<int> depths = parseDepths(filePath);

    cout << "Part 1: " + std::to_string(getIncreaseCount(depths)) << endl;
    cout << "Part 2: " + std::to_string(getSlidingWindowIncreaseCount(depths)) << endl;

    return 0;    
}

File: 2021/day1/solution.py

from typing import List


def get_input() -> List[int]:
    with open('input.txt', 'r') as file:
        lines = []
        for line in file:
            lines.append(int(line.strip()))
        
        return lines

def get_increase_count(depths: List[int]) -> int:
    increases = 0
    for i in range(1, len(depths)):
        if depths[i] > depths[i-1]:
            increases += 1
    
    return increases

def get_sliding_increase_count(depths: List[int]) -> int:
    window = depths[0] + depths[1] + depths[2]
    increases = 0
    for i in range(3, len(depths)):
        new_window = window + depths[i] - depths[i-3]
        if new_window > window:
            increases += 1
        window = new_window
    
    return increases

puzzle_input = get_input()

p1 = get_increase_count(puzzle_input)
print(p1)

p2 = get_sliding_increase_count(puzzle_input)
print(p2)
